# Introduction


[TensorFlow](https://www.tensorflow.org/) is an open source library for numerical computation, specializing in machine learning applications.
This part aims at using transfer learning technique to classify plankton.

### What you will build

In this codelab, you will learn how to run TensorFlow on a single machine, and will train a simple classifier to classify images of plankton.


![](./assets/60.image.jpg)

![jelly](./assets/results_jelly.png)

We will be using transfer learning, which means we are starting with a model that has been already trained on another problem. We will then retrain it on a similar problem. Deep learning from scratch can take days, but transfer learning can be done in short order.

We are going to use a model trained on the [ImageNet](http://image-net.org/) Large Visual Recognition Challenge [dataset](http://www.image-net.org/challenges/LSVRC/2012/). These models can differentiate between 1,000 different classes, like Dalmatian or dishwasher. You will have a choice of model architectures, so you can determine the right tradeoff between speed, size and accuracy for your problem.

We will use this same model, but retrain it to tell apart a small number of classes based on our own examples.


### What you'll Learn

- How to use Python and TensorFlow to train an image classifier
- How to classify images with your trained classifier

### What you need

- A basic Python knowledge
- A basic understanding of Linux commands

## Plankton

Plankton are critically important to our ecosystem, accounting for more than half the primary productivity on earth and nearly half the total carbon fixed in the global carbon cycle. They form the foundation of aquatic food webs including those of large, important fisheries. Loss of plankton populations could result in ecological upheaval as well as negative societal impacts, particularly in indigenous cultures and the developing world. Plankton’s global significance makes their population levels an ideal measure of the health of the world’s oceans and ecosystems.

![plankton](https://storage.googleapis.com/kaggle-competitions/kaggle/3978/media/Plankton-Diagram3.png)

Traditional methods for measuring and monitoring plankton populations are time consuming and cannot scale to the granularity or scope necessary for large-scale studies. Improved approaches are needed. One such approach is through the use of an underwater imagery sensor. This towed, underwater camera system captures microscopic, high-resolution images over large study areas. The images can then be analyzed to assess species populations and distributions.

Manual analysis of the imagery is infeasible – it would take a year or more to manually analyze the imagery volume captured in a single day. Automated image classification using machine learning tools is an alternative to the manual approach. Analytics will allow analysis at speeds and scales previously thought impossible. The automated system will have broad applications for assessment of ocean and ecosystem health.

## Setup and installation 

### Install TensorFlow

Before we can begin the tutorial you need to [install TensorFlow](https://www.tensorflow.org/versions/r1.7/install/) version 1.7

In [12]:
# Imports
!pip install --upgrade "tensorflow==1.7.*"

  Using cached https://files.pythonhosted.org/packages/d3/90/f73789d9f6ecbd629ef646e234af9d15a6fadf81a928f9ae4332ba85fd76/tensorflow-1.7.1-cp36-cp36m-macosx_10_11_x86_64.whl
    100% |████████████████████████████████| 1.8MB 1.2MB/s ta 0:00:011
  Using cached https://files.pythonhosted.org/packages/35/6b/11530768cac581a12952a2aad00e1526b89d242d0b9f59534ef6e6a1752f/astor-0.7.1-py2.py3-none-any.whl
    100% |████████████████████████████████| 102kB 652kB/s a 0:00:01
  Using cached https://files.pythonhosted.org/packages/0b/ec/65d4e8410038ca2a78c09034094403d231228d0ddcae7d470b223456e55d/tensorboard-1.7.0-py3-none-any.whl
    100% |████████████████████████████████| 1.3MB 1.2MB/s ta 0:00:01
  Using cached https://files.pythonhosted.org/packages/33/70/86c5fec937ea4964184d4d6c4f0b9551564f821e1c3575907639036d9b90/bleach-1.5.0-py2.py3-none-any.whl
    100% |████████████████████████████████| 337kB 263kB/s ta 0:00:01
    100% |████████████████████████████████| 92kB 527kB/s ta 0:00:01
  Stored in di

### Clone the git repository

All the code used in this codelab is contained in this git repository. Clone the repository and cd into it. This is where we will be working.


In [2]:
!git clone https://github.com/aymen-mouelhi/cml-plankton-classifier

Cloning into 'cml-plankton-classifier'...
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (20/20), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 20 (delta 4), reused 20 (delta 4), pack-reused 0
Unpacking objects: 100% (20/20), done.


In [3]:
%cd cml-plankton-classifier

/Users/i060307/Desktop/Projects/cml-plankton/cml-plankton-classifier


In [4]:
!ls

CONTRIBUTING.md LICENSE         README.md       scripts         tf_files


## Download the training images

Before you start any training, you'll need a set of images to teach the model about the new classes you want to recognize. We've created an archive of creative-commons licensed flower photos to use initially. Download the photos (218 MB) by invoking the following two commands:

In [5]:
# Download the training images
!curl https://cmlplankton.s3.amazonaws.com/images.zip \
    | tar xz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  8  791M    8 65.5M    0     0   133k      0  1:41:13  0:08:23  1:32:50     0  0     0   281k      0  0:47:53  0:00:30  0:47:23  219k272k      0  0:49:29  0:00:36  0:48:53  238k 0   264k      0  0:51:07  0:00:45  0:50:22  153k8M    0     0   261k      0  0:51:36  0:00:46  0:50:50  153k   256k      0  0:52:36  0:01:23  0:51:13  242k   248k      0  0:54:20  0:02:07  0:52:13  156k      0  0:54:53  0:02:15  0:52:38  242k0   251k      0  0:53:41  0:03:22  0:50:19  197k252k      0  0:53:22  0:03:32  0:49:50  190k 242k      0  0:55:37  0:04:10  0:51:27  260kM    0     0   231k      0  0:58:14  0:04:38  0:53:36  129kM    0     0   231k      0  0:58:19  0:04:42  0:53:37  204k 8 65.5M    0     0   217k      0  1:01:59  0:05:08  0:56:51     08 65.5M    0     0   214k      0  1:03:00  0:05:13  0:57:47     0 0     0   212k      0  1:03:24  0:

You should now have a copy of the flower photos. Confirm the contents of your working directory by issuing the following command:

In [1]:
!ls images

ls: images: No such file or directory


You should now have something like this: 

![](./assets/folder.png)

<span style="color:blue">Note: You will need to open a new terminal window and copy the instructions below.</span>

##  (Re)training the network

### Configure your MobileNet

In this exercise, we will retrain a [MobileNet](https://research.googleblog.com/2017/06/mobilenets-open-source-models-for.html). MobileNet is a a small efficient convolutional neural network. "Convolutional" just means that the same calculations are performed at each location in the image.

The MobileNet is configurable in two ways:

Input image resolution: 128,160,192, or 224px. Unsurprisingly, feeding in a higher resolution image takes more processing time, but results in better classification accuracy.
The relative size of the model as a fraction of the largest MobileNet: 1.0, 0.75, 0.50, or 0.25.
We will use 224 0.5 for this codelab.

With the recommended settings, it typically takes only a couple of minutes to retrain on a laptop. You will pass the settings inside Linux shell variables. Set those variables in your shell:

In [2]:
IMAGE_SIZE=224

In [ ]:
ARCHITECTURE="mobilenet_0.50_${IMAGE_SIZE}"

### More about MobileNet performance (optional)

The graph below shows the first-choice-accuracies of these configurations (y-axis), vs the number of calculations required (x-axis), and the size of the model (circle area).

16 points are shown for MobileNet. For each of the 4 model sizes (circle area in the figure) there is one point for each image resolution setting. The 128px image size models are represented by the lower-left point in each set, while the 224px models are in the upper right.

Other notable architectures are also included for reference. "GoogleNet" in this figure is "Inception V1" in this table. An extended version of this figure is available in slides 84-89 here.

![models](https://codelabs.developers.google.com/codelabs/tensorflow-for-poets/img/70170cbb89d318b1.png)

### Start TensorBoard

Before starting the training, launch tensorboard in the background. TensorBoard is a monitoring and inspection tool included with tensorflow. You will use it to monitor the training progress.

In [27]:
!tensorboard --logdir tf_files/training_summaries &

2019-04-04 16:18:11.555560: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
TensorBoard 1.7.0 at http://C02X80SCJGH5:6006 (Press CTRL+C to quit)
^C


#### Note:
This command will fail with the following error if you already have a tensorboard process running:

<span style="color:red">ERROR:tensorflow:TensorBoard attempted to bind to port 6006, but it was already in use</span>

You can kill all existing TensorBoard instances with:


In [4]:
# Uncomment to kill all existing TensorBoard instances
# !pkill -f "tensorboard"

#### Investigate the retraining script

The retrain script is from the TensorFlow Hub repo, but it is not installed as part of the pip package. So for simplicity I've included it in the codelab repository. You can run the script using the python command. Take a minute to skim its "help".

In [30]:
!python -m scripts.retrain -h

usage: retrain.py [-h] [--image_dir IMAGE_DIR] [--output_graph OUTPUT_GRAPH]
                  [--intermediate_output_graphs_dir INTERMEDIATE_OUTPUT_GRAPHS_DIR]
                  [--intermediate_store_frequency INTERMEDIATE_STORE_FREQUENCY]
                  [--output_labels OUTPUT_LABELS]
                  [--summaries_dir SUMMARIES_DIR]
                  [--how_many_training_steps HOW_MANY_TRAINING_STEPS]
                  [--learning_rate LEARNING_RATE]
                  [--testing_percentage TESTING_PERCENTAGE]
                  [--validation_percentage VALIDATION_PERCENTAGE]
                  [--eval_step_interval EVAL_STEP_INTERVAL]
                  [--train_batch_size TRAIN_BATCH_SIZE]
                  [--test_batch_size TEST_BATCH_SIZE]
                  [--validation_batch_size VALIDATION_BATCH_SIZE]
                  [--print_misclassified_test_images] [--model_dir MODEL_DIR]
                  [--bottleneck_dir BOTTLENECK_DIR]
                  [--final_tensor_name FINAL_TE

#### Run the training

As noted in the introduction, ImageNet models are networks with millions of parameters that can differentiate a large number of classes. We're only training the final layer of that network, so training will end in a reasonable amount of time.

Start your retraining with one big command (note the --summaries_dir option, sending training progress reports to the directory that tensorboard is monitoring) :

In [13]:
%cd cml-plankton-classifier

/Users/i060307/Desktop/Projects/cml-plankton/cml-plankton-classifier


In [14]:
!pygmentize scripts/retrain.py

# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
r"""Simple transfer learning with Inception v3 or Mobilenet models.

With support for TensorBoard.

This example shows how to take a Inception v3 or Mobilenet model trained on
ImageNet images, and train a new top layer that can recognize other classes of
images.

The top layer receives as input a 2048-dimensio

In [ ]:
!python -m scripts.retrain \
  --bottleneck_dir=tf_files/bottlenecks \
  --how_many_training_steps=500 \
  --model_dir=tf_files/models/ \
  --summaries_dir=tf_files/training_summaries/"${ARCHITECTURE}" \
  --output_graph=tf_files/retrained_graph.pb \
  --output_labels=tf_files/retrained_labels.txt \
  --architecture="${ARCHITECTURE}" \
  --image_dir=images

Note that this step will take a while.

This script downloads the pre-trained model, adds a new final layer, and trains that layer on the plankton photos you've downloaded.

ImageNet does not include any of these plankton species we're training on here. However, the kinds of information that make it possible for ImageNet to differentiate among 1,000 classes are also useful for distinguishing other objects. By using this pre-trained network, we are using that information as input to the final classification layer that distinguishes our plankton classes.

#### Bottleneck

A bottleneck is an informal term Google often use for the layer just before the final output layer that actually does the classification. "Bottleneck" is not used to imply that the layer is slowing down the network. We use the term bottleneck because near the output, the representation is much more compact than in the main body of the network.

Every image is reused multiple times during training. Calculating the layers behind the bottleneck for each image takes a significant amount of time. Since these lower layers of the network are not being modified their outputs can be cached and reused.

#### Optional: I'm NOT in a hurry!
The first retraining command iterates only 500 times. You can very likely get improved results (i.e. higher accuracy) by training for longer. To get this improvement, remove the parameter --how_many_training_steps to use the default 4,000 iterations.

In [ ]:
!python -m scripts.retrain \
  --bottleneck_dir=tf_files/bottlenecks \
  --model_dir=tf_files/models/"${ARCHITECTURE}" \
  --summaries_dir=tf_files/training_summaries/"${ARCHITECTURE}" \
  --output_graph=tf_files/retrained_graph.pb \
  --output_labels=tf_files/retrained_labels.txt \
  --architecture="${ARCHITECTURE}" \
  --image_dir=images

### Training And TensorBoard

Once the script finishes generating all the bottleneck files, the actual training of the final layer of the network begins.

By default, this script runs 4,000 training steps. Each step chooses 10 images at random from the training set, finds their bottlenecks from the cache, and feeds them into the final layer to get predictions. Those predictions are then compared against the actual labels, and the results of this comparison is used to update the final layer's weights through a backpropagation process.

As it trains, you'll see a series of step outputs, each one showing training accuracy, validation accuracy, and the cross entropy:

- The training accuracy shows the percentage of the images used in the current training batch that were labeled with the correct class.
- Validation accuracy: The validation accuracy is the precision (percentage of correctly-labelled images) on a randomly-selected group of images from a different set.
- Cross entropy is a loss function that gives a glimpse into how well the learning process is progressing. (Lower numbers are better.)


The figures below show an example of the progress of the model's accuracy and cross entropy as it trains. If your model has finished generating the bottleneck files you can check your model's progress by [opening TensorBoard](http://0.0.0.0:6006/), and clicking on the figure's name to show them. Ignore any warnings that TensorBoard prints to your command line.

The first figure shows accuracy (y-axis) as a function of training progress (x-axis):

![graph](./assets/train.png)

Two lines are shown. The orange line shows the accuracy of the model on the training data. While the blue line shows the accuracy on the test set (which was not used for training). This is a much better measure of the true performance of the network. If the training accuracy continues to rise while the validation accuracy decreases then the model is said to be "overfitting". Overfitting is when the model begins to memorize the training set instead of understanding general patterns in the data.

As the process continues, you should see the reported accuracy improve. After all the training steps are complete, the script runs a final test accuracy evaluation on a set of images that are kept separate from the training and validation pictures. This test evaluation provides the best estimate of how the trained model will perform on the classification task.

You should see an accuracy value of between 85% and 99%, though the exact value will vary from run to run since there's randomness in the training process. (If you are only training on two classes, you should expect higher accuracy.) This number value indicates the percentage of the images in the test set that are given the correct label after the model is fully trained.

### Using the Retrained Model

The retraining script writes data to the following two files:

- ***tf_files/retrained_graph.pb***, which contains a version of the selected network with a final layer retrained on your categories.
- ***tf_files/retrained_labels.txt***, which is a text file containing labels.

#### Classifying an image

The codelab repo also contains a copy of tensorflow's label_image.py example, which you can use to test your network. Take a minute to read the help for this script:

In [15]:
!python -m scripts.label_image -h

/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)
/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
usage: label_image.py [-h] [--image IMAGE] [--graph GRAPH] [--labels LABELS]
                      [--input_height INPUT_HEIGHT]
                      [--input_width INPUT_WIDTH] [--input_mean INPUT_MEAN]
                      [--input_std INPUT_STD] [--input_layer INPUT_LAYER]
                      [--output_layer OUTPUT_LAYER]

optional arguments:
  -h, --help            show this help message and exit
  --image IMAGE         image to be processed
  --graph GRAPH         graph/

Now, let's run the script on this image of a Calanoida:

![calanoida](./cml-plankton-classifier/images/Calanoida/10.dx5duqtumaam4og.jpg)

In [ ]:
!python -m scripts.label_image \
    --graph=tf_files/retrained_graph.pb  \
    --image=images/Calanoida/10.dx5duqtumaam4og.jpg

Each execution will print a list of plankton labels, in most cases with the correct plankton on top (though each retrained model may be slightly different).

You might get results like this for a Calanoida photo:

![results](./assets/results_Calanoida.png)

This indicates a high confidence (~95%) that the image is a Calanoida, and low confidence for any other label.

You can use label_image.py to classify any image file you choose, either from your downloaded collection, or new ones. You just have to change the --image file name argument to the script.

![another](https://upload.wikimedia.org/wikipedia/commons/thumb/f/f0/Cyanea_kils.jpg/180px-Cyanea_kils.jpg)

In [ ]:
!python -m scripts.label_image \
    --graph=tf_files/retrained_graph.pb  \
    --image=../assets/180px-Cyanea_kils.jpg 

### Deploy our Classification Model

####  Flask

Flask is a micro web framework written in Python. It is classified as a microframework because it does not require particular tools or libraries.[3] It has no database abstraction layer, form validation, or any other components where pre-existing third-party libraries provide common functions. However, Flask supports extensions that can add application features as if they were implemented in Flask itself. Extensions exist for object-relational mappers, form validation, upload handling, various open authentication technologies and several common framework related tools. Extensions are updated far more regularly than the core Flask program.[4]

Applications that use the Flask framework include Pinterest,[5] LinkedIn,[6] and the community web page for Flask itself.

In [ ]:
#### Clone the Github repository
!git clone https://github.com/aymen-mouelhi/cml-plankton-flask.git

## TODO: Display the section that does classify

In [ ]:
#### Copy your model files
cp tf_files/models/mobilenet_0.50_224 ./cml-plankton-flask/models/mobilenet

In [ ]:
cp tf_files/models/retrained_graph.pb ./cml-plankton-flask/models/mobilenet

In [ ]:
cp tf_files/models/retrained_labels.txt ./cml-plankton-flask/models/mobilenet

In [9]:
%cd ..

/Users/i060307/Desktop/Projects/cml-plankton


In [ ]:
%cd cml-plankton-flask

In [ ]:
!pygmentize

In [ ]:
# start web server
!python app.py

### Testing

In order to test our chatbot, we need to deploy it in a webserver where it can be accessible from recast.ai
This can be done by deploying our Flask application to Cloud Foundry or to aws. But, as this require a complicated setup, we can use ngrok instead, for testing purposes only. Ngrok will allow external access to your defined http(s) port and will provide an external url that can be included in recast.


#### Download ngrok
You can get the latest version of ngrok from [here](https://ngrok.com/)

In [ ]:
#### Deploy using ngrok
!ngrok http 5000

## Chatbot

In this part we will start creating our chatbot using SAP Conversational AI. 


### Chatbot Settings

We will need to tell our chatbot from where it will get the answers for the classified plankton. Head to the settings section and insert your HTTPS link for your web server


![configuration](./assets/configuration.png)

### Intents Creation

Our bot will need to have an intent for classifying plankton images. 
Let's start by creating the intent named "classify_intent"

![intent](./assets/intent.png)

Once we have created the intent, the next step is to define what should we do once we detect an intent. This is done throught the tab "Build".
- Create a new skill "handle_classify_intent" and provide a description
- Define when the skill should be trigerred
- As we will be expecting a url from the user, we would need to define a requirement parameter:
    - #url as image
- Now the interesting part, let's define the action. In our case, the chatbot should send an HTTP call to our deployed python application. For this workshop, we won't need authentication, so just :
    - Select the type of the HTTP request: POST
    - type /classify as the endpoint
    
![skill](./assets/skill.png)

## Conclusion

![whale food](./assets/whale_food.png)

That's it ! Throught this workshop, you retrained a deep learning model to classify plankton, and then you used it to asnwer users questions.
The same technique can be used to classify flowers and create a flower clasification chatbot :)